sys/kern/tty.c

```diff
@@ -1147,6 +1149,50 @@ int tty_makedev(devfs_node_t *parent, const char *name, tty_t *tty) {
   return devfs_makedev(parent, name, &tty_vnodeops, tty, &tty->t_vnode);
 }
 
+void tty_set_outq_nonempty_flag(tty_thread_t *ttd, tty_t *tty) {
+  if (ringbuf_empty(&tty->t_outq))
+    ttd->ttd_flags &= ~TTY_THREAD_OUTQ_NONEMPTY;
+  else
+    ttd->ttd_flags |= TTY_THREAD_OUTQ_NONEMPTY;
+}
+
+/* TODO: revisit after per-intr_event ithreads are implemented. */
+static void tty_thread(void *arg) {
+  device_t *dev = arg;
+  uart_state_t *uart = dev->state;
+  tty_t *tty = uart->u_tty;
+  tty_thread_t *ttd = &uart->u_ttd;
+  uint8_t work, byte;
+
+  while (true) {
+    WITH_SPIN_LOCK (&uart->u_lock) {
+      /* Sleep until there's work for us to do. */
+      while ((work = ttd->ttd_flags & TTY_THREAD_WORK_MASK) == 0)
+        cv_wait(&ttd->ttd_cv, &uart->u_lock);
+      ttd->ttd_flags &= ~TTY_THREAD_WORK_MASK;
+    }
+    WITH_MTX_LOCK (&tty->t_lock) {
+      if (work & TTY_THREAD_RXRDY) {
+        /* Move characters from rx_buf into the tty's input queue. */
+        while (uart_getb_lock(uart, &byte))
+          if (!tty_input(tty, byte))
+            klog("dropped character %hhx", byte);
+      }
+      if (work & TTY_THREAD_TXRDY)
+        uart_fill_txbuf(dev);
+    }
+  }
+}
+
+void tty_thread_create(const char *name, device_t *dev) {
```

Comment:
A TTY thread can't be created for any device, so I think it should take `uart_state_t *` instead of `device_t *` as an argument.

Reply:
I agree but then - please correct me if I'm wrong - we can't use driver interfaces. They live in `device_t::driver::interfaces` and I'm not sure if parent reference in `uart_state_t` is a good idea.